### ISSUE TITLE
Tensorflow/Keras Endpoint layer does not work with Tensorflow dataset in eager mode

### ISSUE THREAD

**Issue by soroosh-rzh** [tensorflow/tensorflow#45433]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Google Colab, Mac OS
- TensorFlow version:
2.3.1
- Python version:
3.6.9

**Describe the current behavior**
The Endpoint layer pattern described in Tensorflow documents only works with Numpy arrays and it does not work with Tensorflow dataset
**Describe the expected behavior**
The Endpoint layer is expected to work with Tensorflow dataset as well

**Standalone code to reproduce the issue**
 This code is from Tensorflow documents that works only with Numpy arrays:

```
import tensorflow as tf
from tensorflow import keras
import numpy as np

class LogisticEndpoint(keras.layers.Layer):
    def __init__(self, name=None):
        super(LogisticEndpoint, self).__init__(name=name)
        self.loss_fn = keras.losses.BinaryCrossentropy(from_logits=True)
        self.accuracy_fn = keras.metrics.BinaryAccuracy()

    def call(self, targets, logits, sample_weights=None):
        # Compute the training-time loss value and add it
        # to the layer using `self.add_loss()`.
        loss = self.loss_fn(targets, logits, sample_weights)
        self.add_loss(loss)

        # Log accuracy as a metric and add it
        # to the layer using `self.add_metric()`.
        acc = self.accuracy_fn(targets, logits, sample_weights)
        self.add_metric(acc, name="accuracy")

        # Return the inference-time prediction tensor (for `.predict()`).
        return tf.nn.softmax(logits)

inputs = keras.Input(shape=(3,), name="inputs")
targets = keras.Input(shape=(10,), name="targets")
logits = keras.layers.Dense(10)(inputs)
predictions = LogisticEndpoint(name="predictions")(logits, targets)

model = keras.Model(inputs=[inputs, targets], outputs=predictions)
model.compile(optimizer="adam")

data = {
    "inputs": np.random.random((3, 3)),
    "targets": np.random.random((3, 10)),
}
model.fit(data)
```
**Experiment 1: Using TF dataset in a dictionary**
```
tf_data = {
    'inputs': tf.data.Dataset.from_tensor_slices(np.random.random((3, 3))), 
    'targets': tf.data.Dataset.from_tensor_slices(np.random.random((3, 10)))
}

model.fit(tf_data)
```

![tf1](https://user-images.githubusercontent.com/62025982/101292041-09feef80-37db-11eb-8eb7-2047cde45647.png)

**Experiment 2: TF dataset that contains tuples**
```
data = {
    "inputs": np.random.random((3, 3)),
    "targets": np.random.random((3, 10)),
}
tuple_data = []
for i in range(3):
  tuple_data.append((data['inputs'][i, :], data['targets'][i, :]))

dataset = tf.data.Dataset.from_generator(lambda: tuple_data, (tf.float32, tf.float32))

# to show the dataset content
for x in dataset:
  print(x)

model.fit(dataset)
```
![tf2](https://user-images.githubusercontent.com/62025982/101292452-8397dd00-37dd-11eb-871d-71bd1c113727.png)

**Colab link:**
[https://colab.research.google.com/drive/1zaAgPZ4U49t2PkIl7qdm348ayt7Z5O-0](url)

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.3](https://colab.research.google.com/gist/amahendrakar/c339a19b43ebd9e7a83be67405b2870d/45433.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/1db7335dc8bde1e3e2c23cb44fdfc1a3/45433-tf-nightly.ipynb). Please find the attached gist. Thanks!

**Comment by bhack** [comment]
@amahendrakar Thanks.

@soroosh-rzh 
Can you try with:
```
input = tf.random.uniform([3, 3])
targets = tf.random.uniform([3, 10])
tf_data = tf.data.Dataset.from_tensor_slices({'inputs': input, 'targets': targets})
tf_data = tf_data.batch(2)
model.fit(tf_data)
```

**Comment by soroosh-rzh** [comment]
@bhack 

I tried it, it works! Thanks for the suggestion. The loss is a little bit different from what is gained using numpy but I believe it should be Ok! 
Do you have any idea how it might work without dictionary and using tuples as input/outputs? The real use case for me would be multiple inputs/outputs. I mean something close to what I have mentioned in experiment 2. 
Thanks

**Comment by bhack** [comment]
We use our stack overflow space for this kind of support.
Please submit at https://stackoverflow.com/questions/tagged/tensorflow and close this.

**Comment by soroosh-rzh** [comment]
Correct. Forgetting what would be my use case, I still think the endpoint pattern is not working without a dictionary as input so that may be a still a bug not a question! Thank you for your reply.

**Comment by bhack** [comment]
Can you reformulate your code with your bug hypotesis?

**Comment by rmothukuru** [comment]
@soroosh-rzh,
Can you please respond to the above comment. Thanks!

**Comment by soroosh-rzh** [comment]
@bhack @rmothukuru I will reformulate the code and add a comment over my weekend! Please do not close this issue for now. Thanks

**Comment by rmothukuru** [comment]
@soroosh-rzh,
Can you please respond to the above comment. Thanks!

**Comment by rmothukuru** [comment]
@soroosh-rzh,
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

The issue is resolved by using `tf.data.Dataset.from_tensor_slices` and `tf_data.batch(2)` as mentioned in [bhack's answer](https://github.com/tensorflow/tensorflow/issues/45433#issuecomment-740117545). 

Please raise a new issue with your reformulated code. Thanks!
